feat/3866: document new InfluxDB OSS --hardening-enabled feature

jdstrand opened the pull request:
Closes #3866

Document new InfluxDB OSS --hardening-enabled feature by adding a new page under security/ and between enabling TLS and secrets.

Note, these are suggested changes that you're free to rewrite. I did it against InfluxDB OSS 2.1 since 2.2 hasn't been cut yet. ~~This feature will be backported to 2.1, but that hasn't happened just yet.~~

- [x] Signed the [InfluxData CLA](https://www.influxdata.com/legal/cla/)
  ([if necessary](https://github.com/influxdata/docs-v2/blob/master/CONTRIBUTING.md#sign-the-influxdata-cla))
- [x] Rebased/mergeable

jdstrand commented:
Added @samhld to make sure the language is consistent with Product.

pierwill commented:
Thanks, @jdstrand ! Will review this soon.

sanderson commented on content/influxdb/v2.1/security/enable-hardening.md:
````diff
@@ -0,0 +1,43 @@
+products: [oss]
+---
+
+InfluxDB is used in many environments and depending on your environment, enabling additional security features may be appropriate.
+To enable all security hardening options now and in the future,
+start InfluxDB with the `--hardening-enabled` command line flag:
+
+    ```bash
+    influxd \
+    --hardening-enabled
+    ```
````
```suggestion
InfluxDB {{< current-version >}} provides optional security features that ensure your
InfluxDB instance is secure in whatever environment it's used in.

To enable all [additional security features](#security-features), use the
[`hardening-enabled` configuration option](/influxdb/v2.1/reference/config-options/#hardening-enabled)
when starting InfluxDB.
```

There are three different ways to configure `influxd`; the CLI flag is only one. In the Configuration options doc, we show all three methods for setting config options. This also brings up additional changes that need to be made:

- Add a `hardening-enabled` section to the configuration options doc (`content//influxdb/v2.1/reference/config-options/`). Because this isn't available in 2.1 yet, we need to leave it commented out until 2.1 gets patched. We can port the change to the `oss-2.2-base` branch and uncomment the section.
- Update `data/influxd_flags.yml` with the new flag:

    ```yaml
    - flag: "--flux-log-enabled"
      added: 2.0

    - flag: "--hardening-enabled"
      added: 2.2

    - flag: "-h, --help"
      added: 2.0
      nolink: true
    ```

jdstrand replied:
> * Add a `hardening-enabled` section to the configuration options doc (`content//influxdb/v2.1/reference/config-options/`). Because this isn't available in 2.1 yet, we need to leave it commented out until 2.1 gets patched. We can port the change to the `oss-2.2-base` branch and uncomment the section.

@sanderson - what is the recommended way to comment out things? I read https://github.com/influxdata/docs-v2/blob/master/CONTRIBUTING.md but didn't see anything (maybe I missed it?).

sanderson replied:
@jdstrand Markdown just uses HTML comments `<!--  -->`.

Oh, this reminds me... don't comment out the line item in middle of the the TOC list. It won't work. For now, just add it as a comment below the list with a note to move it into the list once 2.1 gets patched.

samhld commented on content/influxdb/v2.1/security/enable-hardening.md:
````diff
@@ -0,0 +1,43 @@
+To enable all security hardening options now and in the future,
+start InfluxDB with the `--hardening-enabled` command line flag:
+
+    ```bash
````
Just a nit -- is it necessary (or conventional) to multi-line these commands when they're so short?

sanderson commented on content/influxdb/v2.1/security/enable-hardening.md:
```diff
@@ -0,0 +1,43 @@
+
+### Private IP Validation
+
+Certain flux functions (eg, `to()`, `from()`, etc) and [template fetching](/influxdb/v2.1/influxdb-templates/) cause InfluxDB to make HTTP requests over the network.
+When private IP validation is enabled, InfluxDB will first verify that the IP address of the URL is not a private IP address.
+IP addresses are considered private if they fall into one of the following categories:
+* IPv4 loopback (`127.0.0.0/8`)
+* RFC1918 (`10.0.0.0/8`, `172.16.0.0/12`, `192.168.0.0/16`)
+* RFC3927 (`169.254.0.0/16`)
+* IPv6 loopback (`::1/128`)
+* IPv6 link-local (`fe80::/10`)
+* IPv6 unique local (`fc00::/7`)
```
```suggestion
Some Flux functions ([`to()`](/flux/v0.x/stdlib/influxdata/influxdb/to/),
[`from()`](/flux/v0.x/stdlib/influxdata/influxdb/from/), [`http.post()`](/flux/v0.x/stdlib/http/post/), etc.)
and [template fetching](/influxdb/v2.1/influxdb-templates/) can require InfluxDB to make
HTTP requests over the network.
With private IP validation enabled, InfluxDB first verifies that the IP address of the URL is not a private IP address.

IP addresses are considered private if they fall into one of the following categories:

- IPv4 loopback (`127.0.0.0/8`)
- RFC1918 (`10.0.0.0/8`, `172.16.0.0/12`, `192.168.0.0/16`)
- RFC3927 (`169.254.0.0/16`)
- IPv6 loopback (`::1/128`)
- IPv6 link-local (`fe80::/10`)
- IPv6 unique local (`fc00::/7`)
```

sanderson commented on content/influxdb/v2.1/security/enable-hardening.md:
```diff
@@ -0,0 +1,43 @@
+#### Private IP considerations
+If your environment requires that these authenticated HTTP requests be made to private IP addresses,
+omit the use of `--hardening-enabled` and
+consider instead setting up egress firewalling to limit which hosts InfluxDB is allowed to connect.
```
```suggestion
consider instead setting up egress firewalling to limit which hosts InfluxDB is allowed to connect.
{{% /note %}}
```

samhld replied:
Perfect!

sanderson commented:
@jdstrand Here's the content that needs to be added to `content/influxdb/v2.1/reference/configuration-options.md`. Also be sure to add an anchor link to the TOC of config options near the top of the page. All the flags are listed in alphabetical order.

```md
### hardening-enabled
Enable [additional security features](/influxdb/v2.1/security/enable-hardening/)
in InfluxDB.

**Default:** `false`  

| influxd flag          | Environment variable        | Configuration key   |
|:------------          |:--------------------        |:-----------------   |
| `--hardening-enabled` | `INFLUXD_HARDENING_ENABLED` | `hardening-enabled` |

###### influxd flag
```sh
influxd --hardening-enabled
```

###### Environment variable
```sh
export INFLUXD_HARDENING_ENABLED=true
```

###### Configuration file
{{< code-tabs-wrapper >}}
{{% code-tabs %}}
[YAML](#)
[TOML](#)
[JSON](#)
{{% /code-tabs %}}
{{% code-tab-content %}}
```yml
hardening-enabled: true
```
{{% /code-tab-content %}}
{{% code-tab-content %}}
```toml
hardening-enabled = true
```
{{% /code-tab-content %}}
{{% code-tab-content %}}
```json
{
  "hardening-enabled": true
}
```
{{% /code-tab-content %}}
{{< /code-tabs-wrapper >}}
```

jdstrand commented on content/influxdb/v2.1/security/enable-hardening.md:
````diff
@@ -0,0 +1,43 @@
+To enable all security hardening options now and in the future,
+start InfluxDB with the `--hardening-enabled` command line flag:
+
+    ```bash
````
I copied this format from the TLS section, but it is reformatted away now anyway.

jdstrand commented:
@sanderson - I believe I addressed your feedback (I committed your suggestions and added two additional commits). I verified locally it all looks how you want (ie, nothing showing right now, but things look good with `draft: false` and uncommenting the items in the config-options).

timhallinflux commented:
I think this should just land with 2.2 --- and not be part of a 2.1 update.  @samhld to work this through with @jdstrand and the Edge team.

jdstrand commented:
> I think this should just land with 2.2 --- and not be part of a 2.1 update. @samhld to work this through with @jdstrand and the Edge team.

We met a few minutes ago and we will *not* be pursuing this for 2.1. @sanderson - please advise how this PR should change now that it is only targeted for 2.2.

sanderson commented:
@jdstrand it may be easier to close this PR and start a new one base off the `oss-2.2-base` branch. I can do this pretty quick, so don't worry about it. I'll add you as a reviewer. Thanks for putting in all the leg work!

sanderson commented:
Closing in favor of #3873.
